src/shared/popup_list.rs

```diff
@@ -0,0 +1,167 @@
+                )
+                sdf.fill_keep(body)
+
+                sdf.stroke(
+                    border_color,
+                    border_width
+                )
+                return sdf.result
+            }
+        }
+    }
+
+    PopupCloseButton = <Button> {
+        width: Fit,
+        height: Fit,
+        margin: {top: -8}
+
+        draw_icon: {
+            svg_file: (ICO_CLOSE),
+            fn get_color(self) -> vec4 {
+                return #000;
+            }
+        }
+        icon_walk: {width: 12, height: 12}
+    }
+
+    PopupList = {{PopupList}} {
+        width: Fit
+        height: Fit
+        flow: Down
+        popup = <PopupDialog> {
+            close_button = <PopupCloseButton> {}
+        }
+        popup_meta: <PopupDialog> {
+            room_status_label = <Label> {
+                width: 110
+                text: "......"
+                draw_text: {
+                    color: #000
+                }
+            }
+            close_button = <PopupCloseButton> {}
+        }
+    }
+
+}
+
+#[derive(Live, Widget)]
+pub struct PopupList {
+    #[deref]
+    view: View,
+    #[layout]
+    layout: Layout,
+    #[live]
+    popup_meta: Option<LivePtr>,
+    #[rust]
+    popups: Vec<View>,
+    #[rust]
+    popups_data: Vec<String>,
+}
+impl LiveHook for PopupList {
+    fn after_apply(&mut self, cx: &mut Cx, apply: &mut Apply, index: usize, nodes: &[LiveNode]) {
+        for button in self.popups.iter_mut() {
+            if let Some(index) = nodes.child_by_name(index, live_id!(popup_meta).as_field()) {
+                button.apply(cx, apply, index, nodes);
+            }
+        }
+    }
+}
+impl Widget for PopupList {
+    fn handle_event(&mut self, cx: &mut Cx, event: &Event, scope: &mut Scope) {
+        while let Some(message) = POPUP_NOTIFICATION.pop() {
+            self.push(cx, message);
+        }
+        for view in self.popups.iter_mut() {
+            view.handle_event(cx, event, scope);
+        }
+        self.widget_match_event(cx, event, scope);
+    }
+
+    fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
+        let mut data = self.popups_data.iter();
+        cx.begin_turtle(walk, self.layout);
+        for view in self.popups.iter_mut() {
+            if let Some(status) = data.next_back() {
+                view.label(id!(room_status_label)).set_text(&status);
+                let walk = walk.with_margin_bottom(10.0);
+                let _ = view.draw_walk(cx, scope, walk);
+            }
+        }
+        cx.end_turtle();
+        DrawStep::done()
+    }
+}
+impl PopupList {
+    fn push(&mut self, cx: &mut Cx, message: String) {
+        self.popups_data.push(message);
+        let meta = self.popup_meta;
+        if self.popups.len() < self.popups_data.len() {
+            for _ in self.popups.len()..self.popups_data.len() {
+                self.popups.push(View::new_from_ptr(cx, meta));
+            }
+        }
```

Comment:
i tried it in my project & had to add a redraw in case nothing else triggered it.

```suggestion
        }
        cx.redraw_all();
```

Reply:
I'd be very much against this, `cx.redraw_all()` will redraw the entire UI, we should only use that when absolutely necessary.
Can we find to get this working by just redrawing `PopupList`?

Reply:
changed to cx.redraw_area